message.go

```diff
@@ -13,7 +15,31 @@ type rpcMessage interface {
 	DecodeMessage(int, decoder, *protocolHandler, *callContainer) error
 }
 
+type BasicRPCData struct {
+	ctx context.Context
+}
+
+func (r *BasicRPCData) Context() context.Context {
+	if r.ctx == nil {
+		r.ctx = context.Background()
```

Comment:
This is worrisome because Context() may be called from multiple goroutines (or not, but one would have to check the codebase to do so) and so this assignment would be racy. I think it's simpler to just always initialize r.ctx in loadContext to something non-nil.